### PR TITLE
Feature/vertex ai embeddings

### PR DESCRIPTION
**Pull request by Yongtae723**
add component of [vertex ai embeddings](https://js.langchain.com/docs/modules/model_io/models/chat/integrations/google_vertex_ai).

I tested this implementation works well on my local environment.
I am python user and not use JS usually. thus I add some comment before review of you. And also, please review carefully, even outside of the comment section.

Question is about credential.
as described [here](https://js.langchain.com/docs/modules/model_io/models/chat/integrations/google_vertex_ai), Unlike other components, it seems that authentication is taken with a credential file instead of registering an API key.

I prepared credential file on local machine as described [here](https://js.langchain.com/docs/modules/model_io/models/chat/integrations/google_vertex_ai), so that I can test it work.

But I wonder how to do it if I deploy other machine like render.com.
do you have idea about this?

upload credential file or do the same thing via linux?

### CODE REVIEW

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                label: 'Model Name',
+                name: 'modelName',
+                type: 'options',
+                options: [
+                    {
+                        label: 'textembedding-gecko',
+                        name: 'textembedding-gecko'
+                    },
+                    {
+                        label: 'textembedding-gecko@001',
```
Comment (Yongtae723) [inline packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:37]: I guess this model name might be fix version of model like openAI do.
currently, I am contacting with google and after reply coming, I will write content of their reply

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        ]
+    }
+
+    async init(nodeData: INodeData, _: string, options: ICommonObject): Promise<any> {
+        const model = nodeData.inputs?.modelName as string
+        const obj: GoogleVertexAIEmbeddingsParams = {
+            model
+        }
+
+        const embedding = new GoogleVertexAIEmbeddings(obj)
```
[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:52]: only 1 model is supported on embedding, so we dont need any model here - https://js.langchain.com/docs/modules/data_connection/text_embedding/integrations/google_vertex_ai

[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:52]: from the docs, we have to: `You have downloaded the credentials for a service account that is permitted to the project and set the GOOGLE_APPLICATION_CREDENTIALS environment variable to the path of this file.`

and I don't see any where that `GOOGLE_APPLICATION_CREDENTIALS` is set here, so I don't think the implementation here is working?

[the remark this answers is not in the document]
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:52]: >only 1 model is supported on embedding, so we dont need any model here -

Yes. Currently there is only one model, so there is no need to create multiple model names OR there may be no ModelName to begin with.
I code it based on the assumption that model names will be added in the future, but shall I delete it? Or should we make it an additional parameter?

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts (hunk 1 above)
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:37]: I asked google and thire answer is

`@001` is for version settings

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:52]: >and I don't see any where that GOOGLE_APPLICATION_CREDENTIALS is set here, so I don't think the implementation here is working?

as far as my understanding, when we do 
```bash
gcloud auth application-default login
```
application_default_credentials.json will be created from `~/.config/gcloud/application_default_credentials.json` for linux or Mac. And when we didn't set GOOGLE_APPLICATION_CREDENTIALS, `~/.config/gcloud/application_default_credentials.json` will be used as default.

 But I would like to know is this point.
 If we want to deploy flowise on cloud, we have to download or upload credentials.json into cloud machine and set `GOOGLE_APPLICATION_CREDENTIALS ` correctly.
 
 In that case,  I would like to know what is the best way to implementation.

[the remark this answers is not in the document]
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:52]: so how about making a credential for GCP credential.
when user add GCP credential, user can upload credentials.json
then system save credentials.json and its path.

how do feel about this idea?

[the remark this answers is not in the document]
Comment (HenryHengZJ) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:52]: @Yongtae723 here's what I would do - https://github.com/FlowiseAI/Flowise/commit/2f9b0e42bc36255f7a9fbf221df4218895cdd049

Can you try it with your credentials.json file? I tried mine but got 401 permission denied

[the remark this answers is not in the document]
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:70]: Yap!
I did and make succeed!

I will comment on your commit!

**Comment by Yongtae723** [comment]
After you accept this PR, I will modify other PR which related to Vertex AI

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        const location = nodeData.inputs?.location as string
+        const credentialData = await getCredentialData(nodeData.credential ?? '', options)
+        const googleApplicationCredentialFilePath = getCredentialParam('googleApplicationCredentialFilePath', credentialData, nodeData)
+        if (!googleApplicationCredentialFilePath) throw new Error('Please specify your Google Application Credential file path')
+
+        const authOptions: GoogleAuthOptions = {
+            keyFile: googleApplicationCredentialFilePath
+        }
+
+        if (projectID) authOptions.projectId = projectID
```
Comment (HenryHengZJ) [inline packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:64]: do we need this? is it already specified in the credentials.json file?

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+        const obj: GoogleVertexAIEmbeddingsParams = {
+            authOptions
+        }
+        if (location) obj.location = location
```
Comment (HenryHengZJ) [inline packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:69]: same here, is it already specified in the credentials.json file?

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts (hunk 3 above)
Comment (HenryHengZJ) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:64]: if we need those, can we put them in the GoogleAuth.credential.ts, same as how we put `googleApplicationCredentialFilePath`?

**Comment by HenryHengZJ** [comment]
would love a screenshot/video to show how it works

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts (hunk 3 above)
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts:46]: Oh, yes!
I agree with your suggestion.

from my understand, we can use multiple project in single account. And thus, I move into GoogleAuth.credential.ts

File: packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts (hunk 4 above)
Comment (Yongtae723) [reply packages/components/nodes/embeddings/GoogleVertexAIEmbedding/GoogleVertexAIEmbedding.ts]: currently, it seems like only `us-central1` is available on location.
thus I removed.

thank you for suggestion

**Comment by Yongtae723** [comment]
this is the screenshot of latest version.

vertex AI embeddings

**Comment by Yongtae723** [comment]
↑ it is succeeded but we have to upload credential.json to use vertex AI.

if we can upload in flowise interface when we make new credential, this is great.
is that possible technically?

**Comment by HenryHengZJ** [comment]
> ↑ it is succeeded but we have to upload credential.json to use vertex AI.
> 
> if we can upload in flowise interface when we make new credential, this is great. is that possible technically?

from quick search, it seems like we can just pass the whole JSON object inside the `credentials.json` file if we don't want to specify the file path. https://stackoverflow.com/a/72620734

So I would suggest having 2 options for user to fill in the credential. You can either;
- Use `Google Application Credential File Path`
- or `Google Credential Object`

I've updated my code [here](https://github.com/FlowiseAI/Flowise/commit/503b845baa888c0a3158e67216ca9031ef9d66c7). Can you try and see if that works? Appreciate the patience!

**Comment by Yongtae723** [comment]
Oh Nice!! you are truly awesome mate!

thank you! I will test it!

**Comment by Yongtae723** [comment]
hey @HenryHengZJ 
Finally it works with both `path of credential file` and `directly input content of file`! 
I feel so happy!

thank you so much for you patience!
vertex ai embeddings

**Pull request closed without merging**